tremor-script/src/ast/binary.rs

```diff
@@ -0,0 +1,381 @@
+            |v| write_bits(bytes, bits as u8, endianess, buf, used, v as u64),
+        ),
+        BytesDataType::Binary => {
+            if let Some(b) = value.as_bytes().and_then(|b| {
+                if bits == 0 {
+                    Some(b)
+                } else {
+                    b.get(..(bits as usize))
+                }
+            }) {
+                if *used == 0 {
+                    bytes.extend_from_slice(&b);
+                } else {
+                    for v in b {
+                        stry!(write_bits(bytes, 8, endianess, buf, used, *v as u64))
+                    }
+                }
+                Ok(())
+            } else {
+                err("Not a long enough binary", &value)
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::prelude::*;
+    use crate::{registry, Script};
+
+    fn eval_binary(src: &str) -> Vec<u8> {
+        let reg: Registry = registry::registry();
+        // let aggr_reg: AggrRegistry = registry::aggr_registry();
+        let script = Script::parse(&crate::path::load(), "<eval>", src.to_string(), &reg)
+            .expect("failed to compile test script");
+
+        let mut event = Value::object();
+        let mut meta = Value::object();
+        let mut state = Value::null();
+        let value = script
+            .run(
+                &EventContext::new(0, None),
+                AggrType::Emit,
+                &mut event,
+                &mut state,
+                &mut meta,
+            )
+            .expect("failed to run test script");
+        match value {
+            Return::Drop => vec![],
+            Return::Emit { value, .. } => value.as_bytes().unwrap_or_default().to_vec(),
+            Return::EmitEvent { .. } => event.as_bytes().unwrap_or_default().to_vec(),
+        }
+    }
+    #[test]
+    fn test_empty() {
+        let empty: [u8; 0] = [];
+        assert_eq!(eval_binary("<<>>"), empty);
+    }
+
+    #[test]
+    fn test_u4() {
+        assert_eq!(eval_binary("<< 1:4 >>"), [1]);
+        assert_eq!(eval_binary("<< 1:4/little >>"), [1]);
+        assert_eq!(eval_binary("<< 1:4/big >>"), [1]);
+
+        assert_eq!(eval_binary("<< 1:4, 1:4 >>"), [17]);
+        assert_eq!(eval_binary("<< 1:4/little, 1:4/little >>"), [17]);
+        assert_eq!(eval_binary("<< 1:4/big, 1:4/big >>"), [17]);
+    }
+
+    #[test]
+    fn test_u6() {
+        assert_eq!(eval_binary("<< 1:6 >>"), [1]);
+        assert_eq!(eval_binary("<< 1:6/little >>"), [1]);
+        assert_eq!(eval_binary("<< 1:6/big >>"), [1]);
+
+        assert_eq!(eval_binary("<< 1:6, 1:2 >>"), [5]);
+        assert_eq!(eval_binary("<< 1:6/little, 1:2/little >>"), [5]);
+        assert_eq!(eval_binary("<< 1:6/big, 1:2/big >>"), [5]);
+
+        assert_eq!(eval_binary("<< 1:2, 1:6 >>"), [65]);
+        assert_eq!(eval_binary("<< 1:2/little, 1:6/little >>"), [65]);
+        assert_eq!(eval_binary("<< 1:2/big, 1:6/big >>"), [65]);
```

Comment:
is this legal: `<<1:2, 1:2>>` ? wouldnt produce a full byte.
If we have less than a whole byte, we could fill the more significant bits with zeroes.
If we have a tail that doesnt make up a full byte, we could fill the less significant bits with zeroes.

Reply:
:+1: I added some tests for this, it's also documented in the RFC :)